Doc/reference/compound_stmts.rst

```diff
@@ -1490,6 +1493,97 @@ can be used to create instance variables with different implementation details.
+layout conflict. We'll discuss each of these in turn.
+
+First, all base classes must allow subclassing. While most classes allow subclassing,
+some built-in classes do not, such as :class:`bool`::
+
+   class SubBool(bool):  # TypeError
+       pass
+
+To create a consistent MRO, all bases appear in the order
+they were specified in the base class list and every child class must appear before its
+base classes. Below is an example where this fails::
+
+   class Base: pass
+   class Child(Base): pass
+   class Grandchild(Base, Child): pass  # TypeError
+
+In the MRO of ``Grandchild``, ``Child`` must appear before ``Base`` because it is first
+in the base class list, but it must also appear after ``Base`` because it is a child of
+``Base``. This is a contradiction, so the class cannot be defined.
+
+If some of the bases have a custom :term:`metaclass`, the metaclass of the resulting class
+is chosen among the metaclasses of the bases. It must be a metaclass that is a subclass of
+all other candidate metaclasses. If no such metaclass exists, the class cannot be created,
+as explained in :ref:`metaclass-determination`.
+
+Finally, the memory layouts of the bases must be compatible. This means that it must be
+possible to compute a *solid base* for the class. A class is a solid base if it has a
+nonempty :attr:`~object.__slots__` definition; some other classes may also be solid bases,
+depending on the Python implementation.
+
+.. impl-detail::
+
+   In CPython, many but not all classes defined in C are solid bases, including most
+   builtins but excluding most concrete :class:`Exception` classes. Generally, a C class
+   is a solid base if its underlying struct is different in size from its base class.
+
+Every class has a solid base. :class:`object`, the base class, has itself as its solid base.
+If there is a single base, the child class's solid base is that class if it is a solid baes,
+or else the base class's solid base. If there are multiple bases, we first find the solid base
+for each base class to produce a list of candidate solid bases. If there is a unique solid base
+that is a subclass of all others, then that class is the solid base. Otherwise, class creation
+fails.
+
+Example::
+
+   class Solid1:
+      __slots__ = ("solid1",)
+
+   class Solid2:
+      __slots__ = ("solid2",)
+
+   class SolidChild(Solid1):
+      __slots__ = ("solid_child",)
+
+   class C1:  # solid base is `object`
+      pass
+
+   # OK: solid bases are `Solid1` and `object`, and `Solid1` is a subclass of `object`.
+   class C2(Solid1, C1):  # solid base is `Solid1`
+      pass
+
+   # OK: solid bases are `SolidChild` and `Solid1`, and `SolidChild` is a subclass of `Solid1`.
+   class C3(SolidChild, Solid1):  # solid base is `SolidChild`
+      pass
+
+   # Error: solid bases are `Solid1` and `Solid2`, but they are not subclasses of each other.
+   class C4(Solid1, Solid2):  # error: no single solid base
```

Comment:
Let's maybe show the real error here? So it would be grepable: 

```python
Traceback (most recent call last):
  File "<python-input-4>", line 1, in <module>
    class C4(Solid1, Solid2): ...
TypeError: multiple bases have instance lay-out conflict
```

Reply:
I guess then we should also make it a doctest so we'll remember to update it if that error message changes.